lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,712 @@
+        local current_column = column
+        for i = row, math.huge do
+            if type(matrix[i]) ~= "table" then
+                matrix[i] = {}
+            end
+            if next(matrix[i]) == nil then
+                return i, 1
+            end
+            -- Scan columns until the maximum column number
+            for j = current_column, self._private.num_cols do
+                if matrix[i][j] == nil then
+                    return i, j
+                end
+            end
+            current_column = 1
+        end
+    -- Add new widget to next row. Create new column if needed
+    elseif self._private.orientation == "x" then
+        -- Scan cols until the end of time
+        local current_row = row
+        if type(matrix[current_row]) ~= "table" then
+            matrix[current_row] = {}
+        end
+        if next(matrix[current_row]) == nil then
+            return current_row, column
+        end
+        for j = column, math.huge do
+            -- Scan row until the maximum row number
+            for i = current_row, self._private.num_rows do
+                if type(matrix[i]) ~= "table" then
+                    matrix[i] = {}
+                end
+
+                if matrix[i][j] == nil then
+                    return i, j
+                end
+            end
+            current_row = 1
+        end
+    end
+    return nil, nil
+end
+
+
+--- Remove a widget from the layout
+--@DOC_wibox_layout_grid_remove_EXAMPLE@
+-- @param number row The row coordinate of the widget to remove
+-- @param number col The column coordinate of the widget to remove
+-- @treturn boolean index If the operation is successful
+function grid:remove_widget_at(row, col)
+    if not row or row < 1 or row > self._private.num_rows then return false end
+    if not col or col < 1 or col > self._private.num_cols then return false end
+
+    local ret = false
+    for index, data in ipairs(self._private.widgets) do
+        if row >= data.row and row < data.row + data.row_span and
+           col >= data.col and col < data.col + data.col_span then
+            matrix_remove(self._private.matrix, data.row, data.col, data.row_span, data.col_span)
+            table.remove(self._private.widgets, index)
+            ret = true
+            break
+        end
+    end
+
+    if ret then
+        -- Recalculate num_rows and num_cols
+        self:get_shape()
+        self:emit_signal("widget::layout_changed")
+    end
+    return ret
+end
+
+
+--- Remove one or more widgets from the layout
+-- @param widget ... Widgets that should be removed (must at least be one)
+-- @treturn boolean index If the operation is successful
+function grid:remove_widgets(...)
+    local args = { ... }
+    local ret = false
+    for _, rem_widget in ipairs(args) do
+        -- loop over the existing widgets
+        for index, data in ipairs(self._private.widgets) do
+            if data.index ~= index then data.index = index end
```

Comment:
What? Why!?!? That seems quite dubious to me. If the index is not trustworthy, how about just dropping it? I know that it is listed as "for debugging", but I cannot imagine a single good debugging reason for it.